sound/hda/hdac_stream.c

```diff
@@ -309,13 +310,12 @@ struct hdac_stream *snd_hdac_stream_assign(struct hdac_bus *bus,
 			res = azx_dev;
 	}
 	if (res) {
-		spin_lock_irq(&bus->reg_lock);
 		res->opened = 1;
 		res->running = 0;
 		res->assigned_key = key;
 		res->substream = substream;
-		spin_unlock_irq(&bus->reg_lock);
 	}
+	spin_unlock_irq(&bus->reg_lock);
```

Comment:
Looking at the code, the only field that affects this code, that is changed in `snd_hdac_stream_release()` is `.opened`, I also assume, that `snd_hdac_stream_release()` can only be called when `.opened == 1`, it wouldn't be called on an unassigned stream, right? So, if `snd_hdac_stream_release()` is called before it is checked in line 302/303, `.opened` will be 0 and it can be used by `snd_hdac_stream_assign()`. If `snd_hdac_stream_release()` is called after the check, `.opened` is still 1 and the stream is skipped, which is also correct. The only potential problem, that I can see, is if `snd_hdac_stream_release()` and `snd_hdac_stream_assign()` run on different cores, then `snd_hdac_stream_release()` sets `.opened` to 0 before it is checked in line 302/303 above. Then even if `snd_hdac_stream_assign()` is running on a faster core, so that it continues to line 312/- while `snd_hdac_stream_release()` is still between lines 334 and 335 below, the spin-lock will force `snd_hdac_stream_assign()` to wait for `snd_hdac_stream_release()` to complete, so `snd_hdac_stream_assign()` will anyway produce a correct result. So, I don't see a problem there either. Could you describe the race that this patch is fixing?